### ISSUE TITLE
Configuration design: Rename data fields

### ISSUE THREAD

**Issue by malex984** [hilbert/hilbert-cli#20]
Note: this is a separation of super-issue hilbert/hilbert-cli#13

**ServicesAndApplications: 2.2** `label` should be `name`, since it's the name of the app. A label is something you write a name (or something else) on, not a name.

**Profile: 3.1** Same thing with labels

**Profile: 3.2** The opposite of back(ground) is fore(ground), not "front"... but we should use more expressive names. If I understand correctly under "back" you listed "services".

**Station: 4.4** Instead of `ip` we could have something like `address` that can be loaded with an IP or a hostname. Usually libraries take either type of address and handle them transparently.

**Comment by porst17** [comment]
# General

In my opinion, all field names should be descriptive and therefore, I would avoid acronyms for the most part and use actual words. E.g. use `description` instead of `desc`, `password` instead of `pw`. Acronyms which are used in everyday language should be OK, e.g. I'd prefer `ip` over `internet_protocol_address`.

# Names of services, apps and profiles

```
applications:
  app1:
    name: "Beautiful application"
    description: "This a beautiful application."
  app2:
    name: "Ugly application"
    description: "This an ugly application."
```

The IDs `app1` and `app2` are used to reference the applications. `name` and `description` will be shown in the gui or cli.

# Profile 3.2

Change `front` to `application` and `back` to `services`.

# Other

If have no good idea regarding the IP address/host name. @elondaits Which terms do other systems use to configure the connection to a server?

**Comment by malex984** [comment]
**Alex said:**

All `label`'s (and `icon`'s) are meant to be shown by Dashboard only. Field names can be easily changed.

>    _3.2_ The opposite of back(ground) is fore(ground), not "front"... but
>    we should use more expressive names. If I understand correctly under "back"
>    you listed "services" (.

Now it is `back` and `apps`.  Can be easily changed.

**Comment by malex984** [comment]
**Eric said:**

> All `label`'s (and `icon`'s) are meant to be shown by Dashboard only. Field names can be easily changed.

I'm suggesting it now so we can easily change it. After we coded it it's more difficult to change and once we released it's VERY hard.

**Comment by malex984** [comment]
**Alex said:**

> - _4.4_ Instead of ip we could have something like address that can be
>   loaded with an IP or a hostname. Usually libraries take either type of
>   address and handle them transparently.

For some types of hosts already `id` is a DNS host name for others there is no DNS name at all. See also 4.1.

**Comment by malex984** [comment]
**Eric said:**

> For some types of hosts already `id` is a DNS host name for others there is no DNS name at all. 

It doesn't matter if the ID is initially the hostname (although you might want shorter IDs without the full domain)... if there is a single `address` field then you can use that, be it a hostname, an ip address or something else to connect to the server via ssh. With an `ip` field instead you can't specify a hostname, just an IP. It's OK if the ID is set to a hostname, but if the IT department has to change all hostnames (I've seen this happen many times in companies, for a number of reasons) it's not necessary to change the IDs, presets, etc.

My point is simply that a single address field could hold either hostname or IP, and that I can't think of a reason where we need the IP and can't resolve it through the DNS if we don't have it.

**Comment by elondaits** [comment]
I supported `address` instead of `ip` or `hostname` because typically all libraries support using either, so it makes no sense to force the user to input either an ip or a hostname when he could choose one transparently.

Most configurations don't use "address" but rather what the address is for. One option is to use "host" or "host_address"... but since it's "the address of the station" and it's placed inside the station's config section I don't think it's wrong, plus adding the concept of "host" just for that is extra confusion for the user.

**Comment by malex984** [comment]
@porst17 Are you OK with using `services` multiple times as in the following?

```
services:
  serv_id: {}

profiles:
  prof_id: { services: [ serv_id, ...], ... }

stations:
  station_id: { profile: prof_id, default_application: app_id, ... }

stationGroups:
  grp_id: 'profile_id, station_id'

applications:
  app_id: { ... }
```

**Comment by porst17** [comment]
Yes, of course!

**Comment by malex984** [comment]
Ok, it is now reflected in the YAML sample https://gist.github.com/malex984/59d2b62c5098eb1005100f9249719505#file-mini_sample-yml
and in the google document (sse Configuration Design Diagram)

**Comment by porst17** [comment]
I suggest changing `options` of applications and services to `runtime` because the field is reflecting what @malex984 called a `ServiceRuntime`. Instead, `options` can mean anything.

**Comment by elondaits** [comment]
I don't think those are runtimes, because runtime is something else and provides other types of services... but at least, "runtime" is a single word so it should be camel cased as ServiceRuntime and not ServiceRunTime as it is now.

**Comment by porst17** [comment]
@elondaits What do you suggest? `options` is clearly not optimal. `runtime` seems too high level. I can also think of `driver`, but this seems to low-level ...

**Comment by malex984** [comment]
Updated (i used `runtime_settings`). See https://github.com/hilbert/hilbert-cli/commit/fece7b8130d5f2c48e53bdac70fa86b7779b824b

**Comment by elondaits** [comment]
@porst17 Actually the presence of runtime_settings seems to be a need because of the idea to model the configuration using a class model, which requires adding class composition because inheritance is being used for something else... but not a problem we'd really have in a YAML config which doesn't really use inheritance or composition (YAML files are formalized in schemas rather, but that's not necessary either).

If, say, "docker-compose" requires a file uri, but "exec" requires a path we could have:

```
Applications:
  earth_from_mars:
    name:  "Earth from Mars"
    type: "docker-compose"
    uri: "..."
  crystals:
    name: "Crystals"
    type: "exec"
    path: "c:\apps\crystals.exe"
```

This is the most normal thing in a cfg. Having an extra level of nesting doesn't make it easier to manage or read.

If "type" describes the "type of application" being described then the current "runtime_settings" are actually the application configuration... so I would put them under the application configuration, like above.

**Comment by malex984** [comment]
@elondaits `runtime_settings` (`service_settings`) is now only for clear class structure. Its content is embedded into `Service`, which is now a Variadic type: depending on its `type` value it may contain different fields. Same was also done with `PowerOnMethod`. 
Note: `Application` is basically the same as `Service` since it extends `Service` with a few extra fields: `name`, `description`, `icon`, `compatibleStations` therefore these fields (alongside with the shared `type`, `auto_detections`) should not appear in any of `ServiceSettings` implementations (e.g. `Compose` should not have `name`)

Note: `ApplicationID` is just a `ServiceID` - they should not be equal.

**Comment by malex984** [comment]
About Services/Applications see (#20)

**Comment by malex984** [comment]
@porst17 @elondaits Q: for consistent naming i would like to propose the following renaming:
 - in a `Group` definition: `intersectWith` -> `intersect_with`
 - in an `Application` definition: `compatibleStations` -> `compatible_stations`


NOTE: according to the `Proposed configuration data model` in the design document - no other data fields are named using camelCase!

**Comment by elondaits** [comment]
Ok!

**Comment by malex984** [comment]
Implemented in git commit 5701b93495d54b1fd0e4dc5ca4ce7d7769082862
within new [hilbert_cli_tools branch](https://github.com/hilbert/hilbert-cli/commits/feature/hilbert_cli_tools)

**Comment by malex984** [comment]
All field names are properly & consistently named now.
